### ISSUE TITLE
Suggestion - automatically query to fill up a limit on filtered results 

### ISSUE THREAD

**Issue by RodH257** [clarkie/dynogels#29]
Hi! First of all, love your work keeping this project alive - Vogels is great but it sucks that it's not so active anymore.

I have a suggestion based on this behaviour:

> Batch Get Items
> 
> Model.getItems allows you to load multiple models with a single request to DynamoDB.
> 
> DynamoDB limits the number of items you can get to 100 or 1MB of data for a single request. Vogels automatically handles splitting up into multiple requests to load all items.

Given that we do something like that, perhaps we can automatically do something like this:

If I query with a 'limit' and then filter the results, the way DynamoDB works is it will query to the limit, and then apply the filter, so take this example:
```  Document
    .query('partitionValue')
    .usingIndex('index')
    .where('sortKey')
    .gte('sortValue')
    .limit(pageSize)
    .filter('un_indexed_field')
    .equals('un_indexed_field_value');
```

Lets say I have 100 items in that table with that partition key, the first 50 all don't have that `un_index_field_value` set for `un_index_field` so they won't show up in the results. That means this query will return 0 items, but will have an indicator that there are more pages.

Basically what we could do, is automatically detect that we:
- Got less than the `limit` value of results
- Got a pointer that there are more pages to request

and then we could automatically request results of `limit - amount_returned` length, and continue until we get limit results or no pointer.

This could be an optional behaviour in the config. It's a fairly minor thing, but thought I'd suggest it!

**Comment by clarkie** [comment]
@RodH257 thanks for your feedback, it's really good to hear there's support out there for this module.

I understand what you're saying but Dynogels handles the limit in the same way the underlying database handles it. I will try and explain why your suggested solution isn't practical but the concept might have legs.

If you have 1000 records in your db and dynamo you want to get 10 with a filter. If the first request returns 9, you will then call subsequent requests with a limit of 1. If the last entry returned is the 10th you'll have done a full table scan in a very inefficient manner. 

You're not the only one to be asking about this sort of thing, see #12 

I'm still mulling over if there is a way to implement this. 

Would an index not solve this problem for you?

**Comment by RodH257** [comment]
@clarkie it could potentially be solved with a clever index, though what I'm doing in my app at the moment is querying on an index, which reduces the amount of data involved and then filtering in 3 different fields within those results, it's hard to make an index that covers all of these fields (and they are optional filters, so it seems unfortunate to waste a GSI on them, maybe a LSI might be more appropriate, though not being able to add them to an existing table sucks). 

Good point about the inefficient table scan (index scan in my case), though my thinking is that in my use case this will end up happening anyway, if the person is using my API with a filter that filters out all but the last one of the 1000 records in your example, and they query for a page size of 10:

- page 1 - 0 results - link to request the next page
- page 2 - 0 results - link to request the next page
and so on until they finally get to the 100th page where it returns that result. 

So they'll end up doing this scan anyway but more inefficiently than me batching it on the server, however this is is possible something thats more specific to my application so it might not make sense to add something generic into the library to solve it.

**Comment by clarkie** [comment]
Unfortunately this is a limitation of dynamodb. I don't think it's worth trying to solve this problem generically here in this module. It's up to your application how you want to store this filtered data. It might be worth using a relational db for the truly dynamic querying / pagination and then just using dynamodb for storing the rest of the data.
